File: frontend/src/types/collaborator.ts
export type CollaboratorRole = 'viewer' | 'editor';
export type CollaboratorStatus = 'pending' | 'accepted' | 'declined';

export interface Collaborator {
  id: number;
  trip_id: number;
  invited_by_user_id: number;
  email: string;
  clerk_user_id: string | null;
  role: CollaboratorRole;
  status: CollaboratorStatus;
  invite_token: string;
  invited_at: string;
  accepted_at: string | null;
}

export interface CollaboratorListResponse {
  trip_id: number;
  collaborators: Collaborator[];
  total: number;
}

export interface InvitePreviewResponse {
  token: string;
  trip_id: number;
  trip_title: string;
  trip_origin: string;
  destinations: string[];
  start_date: string;
  end_date: string;
  invited_by_name: string | null;
  role: CollaboratorRole;
  status: CollaboratorStatus;
}

export interface AcceptInviteResponse {
  success: boolean;
  message: string;
  trip_id: number;
  trip_title: string;
  role: CollaboratorRole;
}

// ── WebSocket / Presence types ──────────────────────────────────────

export interface PresenceUser {
  clerk_id: string;
  display_name: string;
}

// All possible WS message types
export type WSMessageType =
  | 'trip_updated'
  | 'itinerary_generated'
  | 'activity_deleted'
  | 'activity_updated'
  | 'activities_reordered'
  | 'day_replanned'
  | 'collaborator_added'
  | 'collaborator_removed'
  | 'collaborator_role_changed'
  | 'collaborator_joined'
  | 'presence_join'
  | 'presence_leave'
  | 'pong';

export interface WSMessage {
  type: WSMessageType;
  payload: {
    trip_id?: number;
    viewers?: string[];          // display names of active viewers
    clerk_id?: string;
    display_name?: string;
    [key: string]: any;
  };
}

File: backend/app/schemas/collaborator.py
"""
Pydantic schemas for collaborator API
"""
from pydantic import BaseModel, EmailStr, Field, ConfigDict
from typing import Optional, List
from datetime import datetime
from app.models.collaborator import CollaboratorRole, CollaboratorStatus


class InviteCollaboratorRequest(BaseModel):
    email: EmailStr = Field(..., description="Email to invite")
    role: CollaboratorRole = Field(CollaboratorRole.VIEWER, description="viewer or editor")

    model_config = ConfigDict(json_schema_extra={
        "example": {"email": "friend@example.com", "role": "editor"}
    })


class ChangeRoleRequest(BaseModel):
    role: CollaboratorRole = Field(..., description="New role for collaborator")


class CollaboratorResponse(BaseModel):
    id:                 int
    trip_id:            int
    invited_by_user_id: int
    email:              str
    clerk_user_id:      Optional[str]
    role:               CollaboratorRole
    status:             CollaboratorStatus
    invite_token:       str
    invited_at:         datetime
    accepted_at:        Optional[datetime]

    model_config = ConfigDict(from_attributes=True)


class CollaboratorListResponse(BaseModel):
    trip_id:       int
    collaborators: List[CollaboratorResponse]
    total:         int


class InvitePreviewResponse(BaseModel):
    """Returned when validating a token before accepting"""
    token:           str
    trip_id:         int
    trip_title:      str
    trip_origin:     str
    destinations:    List[str]
    start_date:      datetime
    end_date:        datetime
    invited_by_name: Optional[str]
    role:            CollaboratorRole
    status:          CollaboratorStatus


class AcceptInviteResponse(BaseModel):
    success:    bool
    message:    str
    trip_id:    int
    trip_title: str
    role:       CollaboratorRole
